### PR TITLE
Implement non-relocatable OTA-bootloader

### PR DESCRIPTION
**Pull request by kmeshcheryakov**
Implementation of the OTA procedure which uses non-relocatable firmwares

Description
-----------
The OTA procedure uses the simplified scheme for now that allows us to update non-relocatable images. 

ota-statechart_1_001
ota-statechart_1_003

Checklist:
----------

- [x] I have tested my changes. No regression in existing tests.
- [ ] My code is Linted.

### CODE REVIEW

File: demos/nordic/nrf52840-dk/bootloader/main.c [hunk 1]
```diff
@@ -273,29 +289,46 @@ int main( void )
...
             {
+            /* We cannot boot from the second bank for now so we have to commit the image here. 
+            The bootloader can support booting from any position, but the firmware must be made relocatable
+            Unfortunately, it leaves out the case when developer distributed a bad firmware. 
+            
+            So, ideally, we need to get relocatable firmwares to work */
+
                 switch( BANK2_DESCRIPTOR->usImageFlags )
                 {
                     case IMAGE_FLAG_NEW:
```
Comment (huguesBouvier) [inline demos/nordic/nrf52840-dk/bootloader/main.c]: Appart from  vSetBankFlag(0, IMAGE_FLAG_COMMIT_PENDING ); This is a duplicated code from " case IMAGE_FLAG_VALID" can probably be coded in a more sustainable way.

File: demos/nordic/nrf52840-dk/bootloader/main.c [hunk 2]
```diff
@@ -240,21 +246,31 @@ void vCommitSecondBank()
...
-    xErrCode = nrf_bootloader_flash_protect( BOOTLOADER_START_ADDR,
-                                             BOOTLOADER_SIZE, false );
-    APP_ERROR_CHECK( xErrCode );
-
-    /*    uint32_t err_code = sd_softdevice_disable(); */
-    /*    APP_ERROR_CHECK(err_code); */
-
+    vBlinkLeds();
     bool bBank1AFRHeader =
         ( bValidateHeader( ( uint8_t * ) CODE_REGION_1_START ) == NRF_SUCCESS );
```
Comment (huguesBouvier) [inline demos/nordic/nrf52840-dk/bootloader/main.c]: xErrCode = bValidateHeader( ( uint8_t * ) CODE_REGION_2_START );
=> Maybe create a bBank2AFRHeader  for consistency sake

File: demos/nordic/nrf52840-dk/bootloader/main.c [hunk 3]
```diff
@@ -326,23 +359,53 @@ int main( void )
...
+        /* A whole lot of strange things could prevent setting flag to valid, 
+        so we threat CommitPending as a valid state; Anyway, we have already moved the second bank to the first,
+        so we have no choice than to boot 
+        TODO: Consider the commit pending state a failure for the first bank when we implement the relocatable firmwares 
+        */
+        if( xErrCode != NRF_SUCCESS || (BANK1_DESCRIPTOR->usImageFlags != IMAGE_FLAG_VALID &&
+        BANK1_DESCRIPTOR->usImageFlags != IMAGE_FLAG_COMMIT_PENDING )) 
+        /* The first image is corrupted we have no choice but reboot */
         {
             NVIC_SystemReset();
```
Comment (huguesBouvier) [inline demos/nordic/nrf52840-dk/bootloader/main.c]: endless reset cycle?

File: demos/nordic/nrf52840-dk/bootloader/main.c [hunk 4]
```diff
@@ -273,29 +289,46 @@ int main( void )
...
+            So, ideally, we need to get relocatable firmwares to work */
+
                 switch( BANK2_DESCRIPTOR->usImageFlags )
                 {
                     case IMAGE_FLAG_NEW:
-                        vSetSecondBankFlag( IMAGE_FLAG_COMMIT_PENDING );
-                        vBoot( ( uint8_t * ) BANK2_DESCRIPTOR->ulExecutionAddress );
+                        vCommitSecondBank();
+                        xErrCode = xVerifyImageSignature( ( uint8_t * ) CODE_REGION_1_START );
+                        vSetBankFlag(0, IMAGE_FLAG_COMMIT_PENDING );
```
Comment (huguesBouvier) [inline demos/nordic/nrf52840-dk/bootloader/main.c]: Should use a define instead of 0

File: demos/nordic/nrf52840-dk/bootloader/main.c [hunk 5]
```diff
@@ -196,14 +195,21 @@ void vEraseRegion( uint8_t * address,
...
 /** Writes the image flag for the second bank */
-void vSetSecondBankFlag( ImageFlags_t xFlag )
+void vSetBankFlag( uint8_t usBank, ImageFlags_t xFlag )
 {
     ImageDescriptor_t new_desciptor;
-
-    memcpy( &new_desciptor, BANK2_DESCRIPTOR, sizeof( new_desciptor ) );
+    size_t ulCodeRegion = CODE_REGION_1_START;
+    ImageDescriptor_t * pxDescriptor = BANK1_DESCRIPTOR;
+    if (usBank == 1)
```
Comment (huguesBouvier) [inline demos/nordic/nrf52840-dk/bootloader/main.c]: Should use a define instead of 1

File: demos/nordic/nrf52840-dk/bootloader/main.c (hunk 4 above)
Comment (huguesBouvier) [inline demos/nordic/nrf52840-dk/bootloader/main.c]: Set the bank flag after verifying the image.

File: demos/nordic/nrf52840-dk/bootloader/main.c (hunk 2 above)
Comment (kmeshcheryakov) [reply demos/nordic/nrf52840-dk/bootloader/main.c]: Done

File: demos/nordic/nrf52840-dk/bootloader/main.c (hunk 1 above)
Comment (kmeshcheryakov) [reply demos/nordic/nrf52840-dk/bootloader/main.c]: Done

File: demos/nordic/nrf52840-dk/bootloader/main.c (hunk 3 above)
Comment (kmeshcheryakov) [reply demos/nordic/nrf52840-dk/bootloader/main.c]: Changed to endless LED2 blinking

File: demos/nordic/nrf52840-dk/bootloader/main.c (hunk 4 above)
Comment (kmeshcheryakov) [reply demos/nordic/nrf52840-dk/bootloader/main.c]: Done

File: demos/nordic/nrf52840-dk/bootloader/main.c (hunk 5 above)
Comment (kmeshcheryakov) [reply demos/nordic/nrf52840-dk/bootloader/main.c]: Done

File: demos/nordic/nrf52840-dk/bootloader/main.c (hunk 4 above)
Comment (kmeshcheryakov) [reply demos/nordic/nrf52840-dk/bootloader/main.c]: Fixed. That was a major bug.

**Pull request merged**